File: opr_overview/opr_overview/report/opr_production_and_logistics_overview/opr_production_and_logistics_overview.py
import frappe


def execute(filters=None):
	filters = filters or {}

	group_by = filters.get("group_by") or "Project"
	if group_by not in ("Project", "Customer"):
		group_by = "Project"

	view_mode = filters.get("view_mode") or "Logistics"
	if view_mode not in ("Logistics", "Production", "Operations"):
		view_mode = "Logistics"

	columns = get_columns(view_mode)
	data = get_data(filters, group_by)

	return columns, data


def get_data(filters, group_by):

	# =========================================================
	# FILTER VALUES
	# =========================================================

	values = {
		"from_date": filters.get("from_date"),
		"to_date": filters.get("to_date"),

		"project": filters.get("project") or "",
		"project_like": "%" + (filters.get("project") or "") + "%",

		"customer_name": filters.get("customer_name") or "",
		"sales_order": filters.get("sales_order") or "",
		"sales_person": filters.get("sales_person") or "",
		"product_type": filters.get("product_type") or "",
		"region": filters.get("region") or "",
		"workflow_state": filters.get("workflow_state") or "",

		"active_only": 1 if filters.get("active_only") else 0,
		"pending_only": 1 if filters.get("pending_only") else 0,

		"excluded_customer": "Central Ventilation Systems Co. LLC",
	}

	# Sort order follows whichever grouping is selected.
	group_sort_column = "opr.customer_name" if group_by == "Customer" else "opr.project"

	# =========================================================
	# DATABASE QUERY
	#
	# NOTE: "Order Processing Request" has no region column of its
	# own. Region is looked up from the linked Project (matched on
	# opr.project = tabProject.name) since that is the only place
	# a "Region" field exists today.
	# =========================================================

	opr_rows = frappe.db.sql(
		f"""
		SELECT
			opr.name AS opr_name,

			COALESCE(
				NULLIF(TRIM(opr.project), ''),
				'Project Not Specified'
			) AS project,

			COALESCE(
				NULLIF(TRIM(opr.customer_name), ''),
				'Customer Not Specified'
			) AS customer_name,

			COALESCE(
				NULLIF(TRIM(proj.region), ''),
				'Not Specified'
			) AS region,

			COALESCE(
				NULLIF(TRIM(opr.workflow_state), ''),
				'Not Specified'
			) AS workflow_state,

			COALESCE(
				NULLIF(TRIM(opr.product_type1), ''),
				'Not Specified'
			) AS product_type,

			opr.committed_end_date AS committed_end_date,

			COALESCE(opr.total_sqm, 0)
				AS total_sqm,

			COALESCE(opr.total_no, 0)
				AS total_no,

			COALESCE(opr.total_straight_sqm, 0)
				AS total_straight_sqm,

			COALESCE(opr.total_straight_nos, 0)
				AS total_straight_nos,

			COALESCE(opr.total_fittings_sqm, 0)
				AS total_fittings_sqm,

			COALESCE(opr.total_fittings_nos, 0)
				AS total_fittings_nos,

			COALESCE(opr.total_sqm_produced, 0)
				AS total_sqm_produced,

			COALESCE(opr.total_nos_produced, 0)
				AS total_nos_produced,

			COALESCE(opr.remaining_sqm_production, 0)
				AS remaining_sqm_production,

			COALESCE(opr.remaining_nos_production, 0)
				AS remaining_nos_production,

			COALESCE(opr.total_sqm_delivered, 0)
				AS total_sqm_delivered,

			COALESCE(opr.total_nos_delivered, 0)
				AS total_nos_delivered,

			COALESCE(opr.remaining_sqm_delivery, 0)
				AS remaining_sqm_delivery,

			COALESCE(opr.remaining_nos_delivery, 0)
				AS remaining_nos_delivery,

			COALESCE(
				opr.remaining_produced_sqm_to_delivered,
				0
			) AS remaining_produced_sqm_to_delivered,

			COALESCE(
				opr.remaining_produced_no_to_delivered,
				0
			) AS remaining_produced_no_to_delivered

		FROM `tabOrder Processing Request` opr

		LEFT JOIN `tabProject` proj
			ON proj.name = opr.project

		WHERE opr.docstatus < 2

			/* Permanently exclude internal customer */
			AND COALESCE(opr.customer_name, '') <>
				%(excluded_customer)s

			AND (
				%(from_date)s IS NULL
				OR opr.date >= %(from_date)s
			)

			AND (
				%(to_date)s IS NULL
				OR opr.date <= %(to_date)s
			)

			AND (
				%(project)s = ''
				OR COALESCE(opr.project, '') LIKE %(project_like)s
			)

			AND (
				%(customer_name)s = ''
				OR opr.customer_name = %(customer_name)s
			)

			AND (
				%(sales_order)s = ''
				OR opr.sales_order = %(sales_order)s
			)

			AND (
				%(sales_person)s = ''
				OR opr.sales_person = %(sales_person)s
			)

			AND (
				%(product_type)s = ''
				OR opr.product_type1 = %(product_type)s
			)

			AND (
				%(region)s = ''
				OR COALESCE(proj.region, '') = %(region)s
			)

			AND (
				%(workflow_state)s = ''
				OR COALESCE(opr.workflow_state, '') = %(workflow_state)s
			)

			AND (
				%(active_only)s = 0
				OR COALESCE(opr.workflow_state, '') NOT IN (
					'Completed',
					'Canceled',
					'Cancelled',
					'Draft',
					'Submitted for Approval',
					'Approved by General Manager',
					'On Hold',
					'Delivery Completed',
					'PO Verified'
				)
			)

			/*
			 * Pending filter:
			 * - SQM fields intentionally excluded
			 * - Checks remaining Nos and Remaining Value
			 */
			AND (
				%(pending_only)s = 0
				OR (
					COALESCE(
						opr.remaining_nos_production,
						0
					) <> 0

					OR COALESCE(
						opr.remaining_nos_delivery,
						0
					) <> 0

					OR COALESCE(
						opr.remaining_produced_no_to_delivered,
						0
					) <> 0
				)
			)

		ORDER BY
			COALESCE(NULLIF(TRIM({group_sort_column}), ''), 'Not Specified') ASC,
			opr.name ASC
		""",
		values,
		as_dict=True,
	)

	# =========================================================
	# FIELDS TO TOTAL
	# =========================================================

	numeric_fields = [
		"total_sqm",
		"total_no",
		"total_straight_sqm",
		"total_straight_nos",
		"total_fittings_sqm",
		"total_fittings_nos",
		"total_sqm_produced",
		"total_nos_produced",
		"remaining_sqm_production",
		"remaining_nos_production",
		"total_sqm_delivered",
		"total_nos_delivered",
		"remaining_sqm_delivery",
		"remaining_nos_delivery",
		"remaining_produced_sqm_to_delivered",
		"remaining_produced_no_to_delivered",
	]

	# =========================================================
	# GROUP RECORDS (by Project or Customer, per group_by)
	# =========================================================

	grouped_rows = {}
	group_order = []

	for opr in opr_rows:

		if group_by == "Customer":
			group_name = opr.get("customer_name") or "Customer Not Specified"
		else:
			group_name = opr.get("project") or "Project Not Specified"

		if group_name not in grouped_rows:
			grouped_rows[group_name] = []
			group_order.append(group_name)

		grouped_rows[group_name].append(opr)

	# =========================================================
	# INITIALISE TOTALS
	# =========================================================

	grand_total = {fieldname: 0.0 for fieldname in numeric_fields}

	rows = []

	# =========================================================
	# BUILD GROUP SUMMARY ROW + OPR DETAIL ROWS
	# =========================================================

	for group_name in group_order:

		group_total = {fieldname: 0.0 for fieldname in numeric_fields}

		group_oprs = grouped_rows.get(group_name) or []

		group_regions = []
		group_workflow_states = []

		# -----------------------------------------------------
		# CALCULATE GROUP TOTALS
		# -----------------------------------------------------

		for opr in group_oprs:

			region_name = opr.get("region") or "Not Specified"
			workflow_name = opr.get("workflow_state") or "Not Specified"

			if region_name not in group_regions:
				group_regions.append(region_name)

			if workflow_name not in group_workflow_states:
				group_workflow_states.append(workflow_name)

			for fieldname in numeric_fields:

				value = opr.get(fieldname) or 0.0

				group_total[fieldname] = group_total.get(fieldname, 0.0) + value
				grand_total[fieldname] = grand_total.get(fieldname, 0.0) + value

		# -----------------------------------------------------
		# DETERMINE GROUP SUMMARY DISPLAY VALUES
		# -----------------------------------------------------

		group_region = ""
		if len(group_regions) == 1:
			group_region = group_regions[0]
		if len(group_regions) > 1:
			group_region = "Multiple"

		group_workflow_state = ""
		if len(group_workflow_states) == 1:
			group_workflow_state = group_workflow_states[0]
		if len(group_workflow_states) > 1:
			group_workflow_state = "Multiple"

		# -----------------------------------------------------
		# GROUP SUMMARY ROW
		# Customer and Project are always separate columns now;
		# only the field matching the active "Group By" carries
		# the group name, the other stays blank on this row.
		# -----------------------------------------------------

		group_row = {
			"customer_name": group_name if group_by == "Customer" else "",
			"project": group_name if group_by == "Project" else "",
			"region": group_region,
			"workflow_state": group_workflow_state,
			"opr_name": "",
			"committed_end_date": None,
			"product_type": "",
			"indent": 0,
			"row_type": "group_total",
		}

		for fieldname in numeric_fields:
			group_row[fieldname] = group_total.get(fieldname, 0.0)

		rows.append(group_row)

		# -----------------------------------------------------
		# OPR DETAIL ROWS
		# -----------------------------------------------------

		for opr in group_oprs:

			detail_row = {
				"customer_name": opr.get("customer_name"),
				"project": opr.get("project"),
				"region": opr.get("region"),
				"workflow_state": opr.get("workflow_state"),
				"opr_name": opr.get("opr_name"),
				"committed_end_date": opr.get("committed_end_date"),
				"product_type": opr.get("product_type"),
				"indent": 1,
				"row_type": "opr",
			}

			for fieldname in numeric_fields:
				detail_row[fieldname] = opr.get(fieldname) or 0.0

			rows.append(detail_row)

	# =========================================================
	# GRAND TOTAL
	# =========================================================

	grand_total_row = {
		"customer_name": "Grand Total",
		"project": "",
		"region": "",
		"workflow_state": "",
		"opr_name": "",
		"committed_end_date": None,
		"product_type": "",
		"indent": 0,
		"row_type": "grand_total",
	}

	for fieldname in numeric_fields:
		grand_total_row[fieldname] = grand_total.get(fieldname, 0.0)

	rows.append(grand_total_row)

	return rows


def get_columns(view_mode):

	customer_column = {"label": "Customer", "fieldname": "customer_name", "fieldtype": "Data", "width": 200}
	project_column = {"label": "Project", "fieldname": "project", "fieldtype": "Data", "width": 200}
	region_column = {"label": "Region", "fieldname": "region", "fieldtype": "Data", "width": 120}
	opr_no_column = {
		"label": "OPR No",
		"fieldname": "opr_name",
		"fieldtype": "Link",
		"options": "Order Processing Request",
		"width": 135,
	}
	product_type_column = {"label": "Product Type", "fieldname": "product_type", "fieldtype": "Data", "width": 200}
	expected_completion_column = {
		"label": "Exp. Completion Date",
		"fieldname": "committed_end_date",
		"fieldtype": "Date",
		"width": 130,
	}
	workflow_state_column = {"label": "Workflow State", "fieldname": "workflow_state", "fieldtype": "Data", "width": 145}

	opr_qty_columns = [
		{"label": "Total OPR SQM", "fieldname": "total_sqm", "fieldtype": "Float", "precision": 2, "width": 110},
		{"label": "Total OPR Nos", "fieldname": "total_no", "fieldtype": "Float", "precision": 2, "width": 110},
	]
	std_columns = [
		{"label": "Total Std SQM", "fieldname": "total_straight_sqm", "fieldtype": "Float", "precision": 2, "width": 110},
		{"label": "Total Std Nos", "fieldname": "total_straight_nos", "fieldtype": "Float", "precision": 2, "width": 110},
	]
	fitt_columns = [
		{"label": "Total Fitt SQM", "fieldname": "total_fittings_sqm", "fieldtype": "Float", "precision": 2, "width": 110},
		{"label": "Total Fitt Nos", "fieldname": "total_fittings_nos", "fieldtype": "Float", "precision": 2, "width": 110},
	]
	produced_columns = [
		{"label": "Produced SQM", "fieldname": "total_sqm_produced", "fieldtype": "Float", "precision": 2, "width": 110},
		{"label": "Produced Nos", "fieldname": "total_nos_produced", "fieldtype": "Float", "precision": 2, "width": 110},
	]
	rem_production_columns = [
		{
			"label": "Rem. Production SQM",
			"fieldname": "remaining_sqm_production",
			"fieldtype": "Float",
			"precision": 2,
			"width": 135,
		},
		{
			"label": "Rem. Production Nos",
			"fieldname": "remaining_nos_production",
			"fieldtype": "Float",
			"precision": 2,
			"width": 135,
		},
	]
	delivered_columns = [
		{"label": "Delivered SQM", "fieldname": "total_sqm_delivered", "fieldtype": "Float", "precision": 2, "width": 110},
		{"label": "Delivered Nos", "fieldname": "total_nos_delivered", "fieldtype": "Float", "precision": 2, "width": 110},
	]
	rem_delivery_columns = [
		{
			"label": "Rem. Delivery SQM",
			"fieldname": "remaining_sqm_delivery",
			"fieldtype": "Float",
			"precision": 2,
			"width": 125,
		},
		{
			"label": "Rem. Delivery Nos",
			"fieldname": "remaining_nos_delivery",
			"fieldtype": "Float",
			"precision": 2,
			"width": 125,
		},
	]
	ready_dispatch_columns = [
		{
			"label": "Ready Dispatch SQM",
			"fieldname": "remaining_produced_sqm_to_delivered",
			"fieldtype": "Float",
			"precision": 2,
			"width": 140,
		},
		{
			"label": "Ready Dispatch Nos",
			"fieldname": "remaining_produced_no_to_delivered",
			"fieldtype": "Float",
			"precision": 2,
			"width": 140,
		},
	]

	if view_mode == "Production":
		return (
			[customer_column, project_column, region_column, opr_no_column, product_type_column, expected_completion_column]
			+ opr_qty_columns
			+ std_columns
			+ fitt_columns
			+ produced_columns
			+ rem_production_columns
			+ [workflow_state_column]
		)

	if view_mode == "Logistics":
		return (
			[customer_column, project_column, region_column, opr_no_column, product_type_column]
			+ opr_qty_columns
			+ delivered_columns
			+ rem_delivery_columns
			+ ready_dispatch_columns
			+ [workflow_state_column]
		)

	# Operations: the full, combined column set.
	return (
		[customer_column, project_column, region_column, opr_no_column, product_type_column]
		+ opr_qty_columns
		+ std_columns
		+ fitt_columns
		+ produced_columns
		+ rem_production_columns
		+ delivered_columns
		+ rem_delivery_columns
		+ ready_dispatch_columns
		+ [workflow_state_column]
	)

File: opr_overview/opr_overview/report/opr_production_and_logistics_overview/opr_production_and_logistics_overview.js
// Column order must match get_columns() in opr_production_and_logistics_overview.py
// Customer and Project are always the first two columns, in that fixed order,
// regardless of the "Group By" filter (which only controls subtotal grouping).
const OPR_FROZEN_FIELDS = ["customer_name", "project"];

// Merged header definitions, one set per "View Mode". Row 1 = broad section
// headers, Row 2 = SQM/Nos pair headers. Field lists must match get_columns()
// for that view exactly (same fields, same order).
const OPR_IDENTITY_FIELDS = ["customer_name", "project", "region", "opr_name"];

const OPR_HEADER_GROUPS = {
	Logistics: {
		top: [
			{ label: "", fields: [...OPR_IDENTITY_FIELDS, "product_type"] },
			{ label: "OPR Info", fields: ["total_sqm", "total_no"] },
			{
				label: "Logistics Info",
				fields: [
					"total_sqm_delivered",
					"total_nos_delivered",
					"remaining_sqm_delivery",
					"remaining_nos_delivery",
					"remaining_produced_sqm_to_delivered",
					"remaining_produced_no_to_delivered",
				],
			},
			{ label: "", fields: ["workflow_state"] },
		],
		sub: [
			{ label: "", fields: [...OPR_IDENTITY_FIELDS, "product_type", "workflow_state"] },
			{ label: "Total OPR Qty", fields: ["total_sqm", "total_no"] },
			{ label: "Delivered Qty", fields: ["total_sqm_delivered", "total_nos_delivered"] },
			{ label: "Balance Delivery Qty", fields: ["remaining_sqm_delivery", "remaining_nos_delivery"] },
			{
				label: "Ready for Dispatch",
				fields: ["remaining_produced_sqm_to_delivered", "remaining_produced_no_to_delivered"],
			},
		],
	},

	Production: {
		top: [
			{ label: "", fields: [...OPR_IDENTITY_FIELDS, "product_type", "committed_end_date"] },
			{
				label: "OPR Info",
				fields: [
					"total_sqm",
					"total_no",
					"total_straight_sqm",
					"total_straight_nos",
					"total_fittings_sqm",
					"total_fittings_nos",
				],
			},
			{
				label: "Manufacturing Info",
				fields: [
					"total_sqm_produced",
					"total_nos_produced",
					"remaining_sqm_production",
					"remaining_nos_production",
				],
			},
			{ label: "", fields: ["workflow_state"] },
		],
		sub: [
			{
				label: "",
				fields: [...OPR_IDENTITY_FIELDS, "product_type", "committed_end_date", "workflow_state"],
			},
			{ label: "Total OPR Qty", fields: ["total_sqm", "total_no"] },
			{ label: "Total STS", fields: ["total_straight_sqm", "total_straight_nos"] },
			{ label: "Total FTS", fields: ["total_fittings_sqm", "total_fittings_nos"] },
			{ label: "Produced Qty", fields: ["total_sqm_produced", "total_nos_produced"] },
			{
				label: "Balance Production Qty",
				fields: ["remaining_sqm_production", "remaining_nos_production"],
			},
		],
	},

	Operations: {
		top: [
			{ label: "", fields: [...OPR_IDENTITY_FIELDS, "product_type"] },
			{
				label: "OPR Info",
				fields: [
					"total_sqm",
					"total_no",
					"total_straight_sqm",
					"total_straight_nos",
					"total_fittings_sqm",
					"total_fittings_nos",
				],
			},
			{
				label: "Manufacturing Info",
				fields: [
					"total_sqm_produced",
					"total_nos_produced",
					"remaining_sqm_production",
					"remaining_nos_production",
				],
			},
			{
				label: "Logistics Info",
				fields: [
					"total_sqm_delivered",
					"total_nos_delivered",
					"remaining_sqm_delivery",
					"remaining_nos_delivery",
					"remaining_produced_sqm_to_delivered",
					"remaining_produced_no_to_delivered",
				],
			},
			{ label: "", fields: ["workflow_state"] },
		],
		sub: [
			{ label: "", fields: [...OPR_IDENTITY_FIELDS, "product_type", "workflow_state"] },
			{ label: "Total OPR Qty", fields: ["total_sqm", "total_no"] },
			{ label: "Total STS", fields: ["total_straight_sqm", "total_straight_nos"] },
			{ label: "Total FTS", fields: ["total_fittings_sqm", "total_fittings_nos"] },
			{ label: "Produced Qty", fields: ["total_sqm_produced", "total_nos_produced"] },
			{
				label: "Balance Production Qty",
				fields: ["remaining_sqm_production", "remaining_nos_production"],
			},
			{ label: "Delivered Qty", fields: ["total_sqm_delivered", "total_nos_delivered"] },
			{
				label: "Balance Delivery Qty",
				fields: ["remaining_sqm_delivery", "remaining_nos_delivery"],
			},
			{
				label: "Ready for Dispatch",
				fields: ["remaining_produced_sqm_to_delivered", "remaining_produced_no_to_delivered"],
			},
		],
	},
};

frappe.query_reports["OPR Production and Logistics Overview"] = {
	filters: [
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Data",
		},
		{
			fieldname: "customer_name",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
		},
		{
			fieldname: "sales_order",
			label: __("Sales Order"),
			fieldtype: "Link",
			options: "Sales Order",
		},
		{
			fieldname: "sales_person",
			label: __("Sales Person"),
			fieldtype: "Link",
			options: "Sales Person",
		},
		{
			fieldname: "product_type",
			label: __("Product Type"),
			fieldtype: "Link",
			options: "Product Type",
		},
		{
			fieldname: "region",
			label: __("Region"),
			fieldtype: "Link",
			options: "Territory",
		},
		{
			fieldname: "workflow_state",
			label: __("Workflow State"),
			fieldtype: "Link",
			options: "Workflow State",
		},
		{
			fieldname: "active_only",
			label: __("Active Only"),
			fieldtype: "Check",
			default: 1,
		},
		{
			fieldname: "pending_only",
			label: __("Pending Only"),
			fieldtype: "Check",
		},
		{
			fieldname: "group_by",
			label: __("Group By"),
			fieldtype: "Select",
			options: ["Project", "Customer"],
			default: "Project",
		},
		{
			fieldname: "view_mode",
			label: __("View Mode"),
			fieldtype: "Select",
			options: ["Logistics", "Production", "Operations"],
			default: "Logistics",
		},
	],

	get_datatable_options(options) {
		// Predictable column indices (Customer = 0, Project = 1, ...) for freezing.
		options.serialNoColumn = false;
		return options;
	},

	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (data && (data.row_type === "group_total" || data.row_type === "grand_total")) {
			value = `<b>${value}</b>`;
		}
		return value;
	},

	after_datatable_render(datatable) {
		const view_mode = frappe.query_report.get_filter_value("view_mode") || "Logistics";
		setup_header_wrap();
		setup_total_row_colour(datatable);
		setup_group_headers(datatable, view_mode);
		setup_freeze_toggle(datatable);
		bind_resize_handler(datatable, view_mode);
	},

	onload(report) {
		report.page.add_inner_button(__("Toggle Freeze Columns"), () => {
			const datatable = report.datatable;
			if (!datatable) return;
			if (datatable._opr_freeze_on) {
				remove_freeze(datatable);
			} else {
				apply_freeze(datatable);
			}
		});
	},
};

function setup_header_wrap() {
	const styleId = "opr-overview-header-wrap-style";
	if (document.getElementById(styleId)) return;

	const style = document.createElement("style");
	style.id = styleId;
	style.textContent = `
		.dt-header .dt-cell {
			height: auto !important;
			min-height: 48px;
			background-color: #dbeeff !important;
		}
		.dt-header .dt-cell__content {
			white-space: normal !important;
			word-break: break-word !important;
			line-height: 1.3 !important;
			height: auto !important;
			min-height: 48px;
			padding-top: 6px !important;
			padding-bottom: 6px !important;
			color: #1a1a1a !important;
		}
		.dt-header .dt-row {
			height: auto !important;
			min-height: 48px;
		}
		.opr-group-header-row .dt-cell {
			background-color: #cfe4fb !important;
		}
		.opr-group-header__content {
			display: flex !important;
			align-items: center;
			justify-content: center;
			text-align: center;
			font-weight: 600;
			color: #1a1a1a !important;
		}
	`;
	document.head.appendChild(style);
}

function setup_total_row_colour(datatable) {
	const rows = datatable.datamanager && datatable.datamanager.data;
	if (!rows) return;

	const group_total_rows = [];
	const grand_total_rows = [];

	rows.forEach((row, i) => {
		if (row.row_type === "group_total") group_total_rows.push(i);
		if (row.row_type === "grand_total") grand_total_rows.push(i);
	});

	if (group_total_rows.length) {
		datatable.style.setStyle(
			group_total_rows.map((i) => `.dt-row-${i}`).join(","),
			{ "background-color": "var(--dt-header-cell-bg)" }
		);
	}

	if (grand_total_rows.length) {
		datatable.style.setStyle(
			grand_total_rows.map((i) => `.dt-row-${i}`).join(","),
			{
				"background-color": "var(--dt-header-cell-bg)",
				"border-top": "2px solid var(--dt-border-color)",
			}
		);
	}
}

function get_field_to_col_index(datatable) {
	const map = {};
	datatable.datamanager.getColumns().forEach((col) => {
		map[col.id] = col.colIndex;
	});
	return map;
}

function get_column_widths(datatable) {
	const widths = [];
	datatable.datamanager.getColumns().forEach((col) => {
		const cell = datatable.header.querySelector(`.dt-cell--header-${col.colIndex}`);
		widths[col.colIndex] = cell ? cell.offsetWidth : col.width || 100;
	});
	return widths;
}

function build_group_row_html(groups, fieldToCol, colWidths, rowClass) {
	const cells = groups.map((group, idx) => {
		const width = group.fields.reduce((sum, fieldname) => {
			const colIndex = fieldToCol[fieldname];
			return sum + (colWidths[colIndex] || 0);
		}, 0);

		const is_frozen_cell = idx === 0;

		return `
			<div class="dt-cell dt-cell--header ${is_frozen_cell ? "opr-group-freeze-cell" : ""}"
				style="width:${width}px;min-width:${width}px;max-width:${width}px;">
				<div class="dt-cell__content opr-group-header__content">${frappe.utils.escape_html(group.label)}</div>
			</div>
		`;
	});

	return `<div class="dt-row ${rowClass}">${cells.join("")}</div>`;
}

function setup_group_headers(datatable, view_mode) {
	// remove any group header rows from a previous render
	datatable.header.querySelectorAll(".opr-group-header-row").forEach((el) => el.remove());

	const groups = OPR_HEADER_GROUPS[view_mode] || OPR_HEADER_GROUPS.Logistics;

	const headerInner = datatable.header.querySelector(":scope > div") || datatable.header;
	const headerRow = headerInner.querySelector(".dt-row-header");
	if (!headerRow) return;

	const fieldToCol = get_field_to_col_index(datatable);
	const colWidths = get_column_widths(datatable);

	const topRowHtml = build_group_row_html(
		groups.top,
		fieldToCol,
		colWidths,
		"opr-group-header-row opr-group-header-row--top"
	);
	const subRowHtml = build_group_row_html(
		groups.sub,
		fieldToCol,
		colWidths,
		"opr-group-header-row opr-group-header-row--sub"
	);

	headerRow.insertAdjacentHTML("beforebegin", topRowHtml + subRowHtml);
}

function apply_freeze(datatable) {
	if (datatable._opr_freeze_on) return;
	datatable._opr_freeze_on = true;

	const fieldToCol = get_field_to_col_index(datatable);
	const frozenIndices = OPR_FROZEN_FIELDS.map((f) => fieldToCol[f]);

	const bodyScrollable = datatable.bodyScrollable;
	const header = datatable.header;
	if (!bodyScrollable || !header) return;

	// Pin the body cells for the frozen columns using sticky positioning
	// (bodyScrollable has real horizontal overflow, so sticky works there).
	const offsets = {};
	let cumLeft = 0;
	frozenIndices.forEach((i) => {
		offsets[i] = cumLeft;
		const cell = header.querySelector(`.dt-cell--header-${i}`);
		cumLeft += cell ? cell.offsetWidth : 0;
	});
	const frozenWidth = cumLeft;

	const styleId = "opr-overview-freeze-style";
	let style = document.getElementById(styleId);
	if (!style) {
		style = document.createElement("style");
		style.id = styleId;
		document.head.appendChild(style);
	}
	style.textContent = frozenIndices
		.map(
			(i) =>
				`.dt-cell--col-${i} { position: sticky !important; left: ${offsets[i]}px !important; z-index: 2 !important; background: var(--dt-cell-bg) !important; }`
		)
		.join("\n");

	// The header (and our injected group-header rows) scroll via a JS-driven
	// translateX on the whole .dt-header, so sticky won't engage there.
	// Counter-translate the frozen header/group cells to keep them visually pinned.
	const onScroll = function () {
		const scrollLeft = this.scrollLeft;

		frozenIndices.forEach((i) => {
			header.querySelectorAll(`.dt-cell--header-${i}`).forEach((cell) => {
				cell.style.transform = `translateX(${scrollLeft}px)`;
				cell.style.position = "relative";
				cell.style.zIndex = "3";
				cell.style.background = "var(--dt-header-cell-bg)";
			});
		});

		header.querySelectorAll(".opr-group-freeze-cell").forEach((cell) => {
			cell.style.transform = `translateX(${scrollLeft}px)`;
			cell.style.position = "relative";
			cell.style.zIndex = "3";
			cell.style.width = `${frozenWidth}px`;
		});
	};

	datatable._opr_freeze_scroll_handler = onScroll;
	bodyScrollable.addEventListener("scroll", onScroll);
	onScroll.call(bodyScrollable);
}

function remove_freeze(datatable) {
	if (!datatable._opr_freeze_on) return;
	datatable._opr_freeze_on = false;

	const style = document.getElementById("opr-overview-freeze-style");
	if (style) style.remove();

	const header = datatable.header;
	if (datatable._opr_freeze_scroll_handler && datatable.bodyScrollable) {
		datatable.bodyScrollable.removeEventListener("scroll", datatable._opr_freeze_scroll_handler);
	}

	header.querySelectorAll('[class*="dt-cell--header-"], .opr-group-freeze-cell').forEach((cell) => {
		cell.style.transform = "";
		cell.style.position = "";
		cell.style.zIndex = "";
	});
}

function bind_resize_handler(datatable, view_mode) {
	if (datatable._opr_resize_bound) return;
	datatable._opr_resize_bound = true;

	window.addEventListener(
		"resize",
		frappe.utils.debounce(() => {
			setup_group_headers(datatable, view_mode);
			if (datatable._opr_freeze_on) {
				remove_freeze(datatable);
				apply_freeze(datatable);
			}
		}, 300)
	);
}

function setup_freeze_toggle(datatable) {
	// Freeze the first 2 columns (Customer, Project) by default; user can
	// toggle it off/on via the "Toggle Freeze Columns" button. Re-applied
	// (not just left as-is) on every render since column widths can change.
	const was_on = datatable._opr_freeze_on !== false;
	if (datatable._opr_freeze_on) remove_freeze(datatable);
	if (was_on) apply_freeze(datatable);
}
